File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_CDKEY2.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.IO;
using System.Linq;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_CDKEY2 : Message
    {
        public const string KEYOWNER_TOOMANYSPAWNS = "TOO MANY SPAWNS";
        public const string KEYOWNER_NOSPAWNING = "NO SPAWNING";

        public enum Statuses : UInt32
        {
            Success = 1,
            InvalidKey = 2,
            BadProduct = 3,
            Banned = 4,
            InUse = 5,
        }

        public SID_CDKEY2()
        {
            Id = (byte)MessageIds.SID_CDKEY2;
            Buffer = new byte[0];
        }

        public SID_CDKEY2(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_CDKEY2;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            switch (context.Direction)
            {
                case MessageDirection.ClientToServer:
                    {
                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, "[" + Common.DirectionToString(context.Direction) + "] SID_CDKEY2 (" + (4 + Buffer.Length) + " bytes)");

                        if (Buffer.Length < 45)
                            throw new GameProtocolViolationException(context.Client, "SID_CDKEY2 must be at least 45 bytes");
                        /**
                         * (UINT32) Spawn Key (1 is TRUE, 0 is FALSE)
                         * (UINT32) Key length
                         * (UINT32) Key Product value
                         * (UINT32) Key Public value
                         * (UINT32) Server Token
                         * (UINT32) Client Token
                         *  (UINT8) [20] Hashed Key Data
                         * (STRING) Key owner name *
                         */

                        var m = new MemoryStream(Buffer);
                        var r = new BinaryReader(m);

                        context.Client.GameState.SpawnKey = (r.ReadUInt32() == 1);
                        var keyLength = r.ReadUInt32();
                        var productValue = r.ReadUInt32();
                        var publicValue = r.ReadUInt32();
                        var serverToken = r.ReadUInt32();
                        context.Client.GameState.ClientToken = r.ReadUInt32();
                        var hashedKeyData = r.ReadBytes(20);
                        context.Client.GameState.KeyOwner = r.ReadString();

                        r.Close();
                        m.Close();

                        if (serverToken != context.Client.GameState.ServerToken)
                            throw new GameProtocolViolationException(context.Client, "SID_CDKEY2 server token mismatch");

                        var gameKey = new GameKey(keyLength, productValue, publicValue, hashedKeyData);
                        context.Client.GameState.GameKeys.Append(gameKey);

                        return new SID_CDKEY2().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient));
                    }
                case MessageDirection.ServerToClient:
                    {
                        /**
                         * (UINT32) Result
                         * (STRING) Key Owner
                         */

                        Buffer = new byte[5];

                        var m = new MemoryStream(Buffer);
                        var w = new BinaryWriter(m);

                        w.Write((UInt32)Statuses.Success);
                        w.Write("");

                        w.Close();
                        m.Close();

                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, "[" + Common.DirectionToString(context.Direction) + "] SID_CDKEY2 (" + (4 + Buffer.Length) + " bytes)");
                        context.Client.Send(ToByteArray());
                        return true;
                    }
            }

            return false;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_CDKEY2.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.IO
Imports System.Linq

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_CDKEY2
        Inherits Message

        Public Const KEYOWNER_TOOMANYSPAWNS As String = "TOO MANY SPAWNS"
        Public Const KEYOWNER_NOSPAWNING As String = "NO SPAWNING"

        Public Enum Statuses As UInt32
            Success = 1
            InvalidKey = 2
            BadProduct = 3
            Banned = 4
            InUse = 5
        End Enum

        Public Sub New()
            Id = CByte(MessageIds.SID_CDKEY2)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_CDKEY2)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Select Case context.Direction
                Case MessageDirection.ClientToServer
                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    If Buffer.Length < 45 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} must be at least 45 bytes")

                    Using m = New MemoryStream(Buffer)
                        Using r = New BinaryReader(m)
                            context.Client.GameState.SpawnKey = (r.ReadUInt32() = 1)
                            Dim keyLength = r.ReadUInt32()
                            Dim productValue = r.ReadUInt32()
                            Dim publicValue = r.ReadUInt32()
                            Dim serverToken = r.ReadUInt32()
                            context.Client.GameState.ClientToken = r.ReadUInt32()
                            Dim hashedKeyData = r.ReadBytes(20)
                            context.Client.GameState.KeyOwner = r.ReadByteString()
                            If serverToken <> context.Client.GameState.ServerToken Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} server token mismatch")
                            Dim gameKey = New GameKey(keyLength, productValue, publicValue, hashedKeyData)
                            Dim unused = context.Client.GameState.GameKeys.Append(gameKey)
                        End Using
                    End Using

                    Return New SID_CDKEY2().Invoke(New MessageContext(context.Client, MessageDirection.ServerToClient))
                Case MessageDirection.ServerToClient
                    Buffer = New Byte(4) {}

                    Using m = New MemoryStream(Buffer)
                        Using w = New BinaryWriter(m)
                            w.Write(CUInt(Statuses.Success))
                            w.Write(CByte(0))
                        End Using
                    End Using

                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    context.Client.Send(ToByteArray(context.Client.ProtocolType))
                    Return True
            End Select

            Return False
        End Function
    End Class
End Namespace
